plutus-ledger-api/test-onchain-evaluation/Main.hs

```diff
@@ -0,0 +1,121 @@
+                                             ScriptEvaluationResult (..))
+import PlutusLedgerApi.V1 qualified as V1
+import PlutusLedgerApi.V2 qualified as V2
+import PyF (fmt)
+import System.Directory.Extra (listFiles)
+import System.Exit (die)
+import System.FilePath (isExtensionOf, (<.>))
+import System.IO.Extra (withTempDir)
+import System.Process.Extra (system_)
+
+data TestFailure
+    = InvalidEvaluationResult
+        Text
+        -- ^ Serialised `ScriptEvaluationEvent`
+        ScriptEvaluationResult
+    | InvalidCostParams
+        Text
+        -- ^ Serialised `ScriptEvaluationEvent`
+        CostModelApplyError
+    deriving stock (Show)
+
+main :: IO ()
+main = do
+    opts <- OA.execParser O.parserInfo
+    let extZipped = O.optsEventFileExt opts <.> "bz2"
+    withTempDir $ \dir -> do
+        putStrLn "Downloading event dump files"
+        let s3DownloadCmd :: String
+            s3DownloadCmd = [fmt|{keyId} {secretKey} {region} aws {endpoint} s3 cp {s3Path} {dir} {options}|]
+              where
+                keyId = [fmt|AWS_ACCESS_KEY_ID={O.optsS3AccessKeyId opts}|]
+                secretKey = [fmt|AWS_SECRET_ACCESS_KEY={O.optsS3SecretAccessKey opts}|]
+                region = [fmt|AWS_DEFAULT_REGION={O.optsS3Region opts}|]
+                endpoint = [fmt|--endpoint-url {O.optsS3Endpoint opts}|]
+                s3Path = [fmt|s3://{O.optsS3Bucket opts}/{O.optsS3Prefix opts}|]
+                options = [fmt|--recursive --exclude "*" --include "*.{extZipped}"|]
+        putStrLn $ "Running: " <> s3DownloadCmd
+        -- It would be better to use amazonka, but the version bounds don't quite line up
+        -- (e.g., amazonka-core has aeson <1.6)
+        system_ s3DownloadCmd
+        eventFilesZipped <- filter (extZipped `isExtensionOf`) <$> listFiles dir
+        for_ eventFilesZipped $ \zipped -> do
+            putStrLn $ "Unzipping " <> zipped
+            system_ $ "bunzip2 " <> zipped
+        eventFiles <- filter (O.optsEventFileExt opts `isExtensionOf`) <$> listFiles dir
+        for_ (zip [1 :: Int ..] eventFiles) $ \(i, eventFile) -> do
+            putStrLn [fmt|Processing event file {i}: {eventFile}|]
+            events <- CBOR.readFileDeserialise @[ScriptEvaluationEvent] eventFile
+            mbErrs <- Concurrent.withPool (O.optsConcurrency opts) $ \pool -> do
+                Concurrent.parallel pool $ fmap (evaluate . checkEvaluationEvent) events
+            case catMaybes mbErrs of
+                [] -> putStrLn "All tests passed"
+                errs ->
+                    die
+                        [fmt|
+Script evaluation regression test failed.
+Number of failed test cases: {length errs}
+{intercalate "\n" (show <$> errs)}
+|]
+
+checkEvaluationEvent :: ScriptEvaluationEvent -> Maybe TestFailure
+checkEvaluationEvent ev = case ev of
+    PlutusV1Event ScriptEvaluationData{..} expected ->
+        case V1.mkEvaluationContext dataCostParams of
```

Comment:
If we group things by epoch, we can share this work, i.e. all events in the same epoch should have the same cost model params. The ledger does that, because it saves a lot of time. That might speed up the test suite a lot.

Potentially if we did that sharing in the dump too we might save a lot of space as wel!

Reply:
> all events in the same epoch should have the same cost model params

That's useful to know. This could significantly reduce the dump size!

Reply:
Also to be extra clear: `mkEvaluationContext` is expensive, so caching its value across the epoch is probably also good.